src/gbinder_fmq.c

```diff
@@ -0,0 +1,794 @@
+            munmap(base_address, map_length);
+        }
+    }
+}
+
+void
+gbinder_fmq_free(
+    GBinderFmq* self)
+{
+    GBinderFmqPriv* priv = gbinder_fmq_cast(self);
+
+    if (self->desc->flags == GBINDER_FMQ_TYPE_UNSYNC_WRITE &&
+            priv->read_ptr != NULL) {
+        g_free(priv->read_ptr);
+    } else if (priv->read_ptr != NULL) {
+        gbinder_fmq_unmap_grantor_descriptor(self, priv->read_ptr,
+            READ_PTR_POS);
+    }
+    if (priv->write_ptr != NULL) {
+        gbinder_fmq_unmap_grantor_descriptor(self, priv->write_ptr,
+            WRITE_PTR_POS);
+    }
+    if (priv->ring != NULL) {
+        gbinder_fmq_unmap_grantor_descriptor(self, priv->ring, DATA_PTR_POS);
+    }
+    if (priv->event_flag_ptr != NULL) {
+        gbinder_fmq_unmap_grantor_descriptor(self, priv->event_flag_ptr,
+            EVENT_FLAG_PTR_POS);
+    }
+
+    if (self->desc->grantors.data.ptr) {
+        g_free((GBinderFmqGrantorDescriptor*)self->desc->grantors.data.ptr);
+    }
+    if (self->desc->data.fds) {
+        g_free((GBinderFds*)self->desc->data.fds);
+    }
+
+    g_free(self->desc);
+    g_slice_free(GBinderFmqPriv, priv);
+}
+
+/* Public API */
+
+GBinderFmq*
+gbinder_fmq_new(
+    gsize item_size,
+    gsize num_items,
+    GBINDER_FMQ_TYPE type,
+    GBINDER_FMQ_FLAGS flags,
+    gint fd,
+    gsize buffer_size)
+{
+    GBinderFmq* self = NULL;
+
+    if (item_size <= 0) {
+        GWARN("Incorrect item size");
+    } else if (num_items <= 0) {
+        GWARN("Empty queue requested");
+    } else if (num_items > SIZE_MAX / item_size) {
+        GWARN("Requested message queue size too large");
+    } else if (fd != -1 && num_items * item_size > buffer_size) {
+        GWARN("The size needed for items (%"G_GSIZE_FORMAT") is larger\
+            than the supplied buffer size (%"G_GSIZE_FORMAT")",
+            num_items * item_size, buffer_size);
+    } else {
+        gboolean configure_event_flag =
+            (flags & GBINDER_FMQ_FLAG_CONFIGURE_EVENT_FLAG) != 0;
+        gsize queue_size_bytes = num_items * item_size;
+        gsize meta_data_size;
+        gsize shmem_size;
+        int shmem_fd;
+        GBinderFmqPriv* priv = g_slice_new0(GBinderFmqPriv);
+
+        self = &priv->pub;
+
+        meta_data_size = 2 * sizeof(guint64);
+        if (configure_event_flag) {
+            meta_data_size += sizeof(guint32);
+        }
+
+        /* Allocate shared memory */
+        if (fd != -1) {
+            /* User-supplied ringbuffer memory provided,
+             * allocating memory only for meta data */
+            shmem_size = (meta_data_size + getpagesize() - 1) &
+                ~(getpagesize() - 1);
+        } else {
+            /* Allocate ringbuffer, read counter and write counter */
+            shmem_size = (G_ALIGN8(queue_size_bytes) +
+                meta_data_size + getpagesize() - 1) & ~(getpagesize() - 1);
+        }
+
+        shmem_fd = memfd_create("MessageQueue", MFD_CLOEXEC);
```

Comment:
On some systems MFD_CLOEXEC requires `#include <linux/memfd.h>`
```
src/gbinder_fmq.c:335:49: error: ‘MFD_CLOEXEC’ undeclared (first use in this function)
         shmem_fd = memfd_create("MessageQueue", MFD_CLOEXEC);
                                                 ^
```

Reply:
in addition to including `<linux/memfd.h>` let's use a direct syscall:
```
        shmem_fd = syscall(__NR_memfd_create, "MessageQueue", MFD_CLOEXEC);
```
glibc wrapper for this system call appeared many years later than the system call itself. Quite a few Linux releases were shipped with memfd_create functionality built into the kernel but without glibc wrapper.

That also allows to drop `#define _GNU_SOURCE 1`, I think.

Reply:
Change to direct syscall and added needed includes.